rfcs/select-component.md

```diff
@@ -0,0 +1,191 @@
+
+### Exports
+
+* `StatefulSelect`
+* `StatefulSelectContainer`
+* `Select`
+* `StyledSelectInput`
+* `StyledOption`
+* `StyledDropDown`
+* `StyledLabel`
+* `StyledHint`
+
+### `Select` and `StatefulSelect` API
+
+* `type: TYPE.SEARCH | TYPE.SELECT`:
+  type of component to be in select or search mode (with lookup input)
+* `selectedOptions: Array<Object>`:
+  Current selected options. Every option object has `id: string` and `label: React$Node`. Label is defaulted to display for selected option, otherwise see `getSelectedOptionLabel` method
+* `placeholder: ?string`:
+  Placeholder text if nothing is selected. Default is `Choose one...`
+* `rows: ?number`:
+  Represents maximum visible length of options, all other will be scrolled. If not defined, all options will be visible.
+* `options: Array<Object>`:
+  All Options in dropdown. Should be provided for Select and Search mode equally. Every option object has `id: string` and `label: React$Node`. Label is defaulted to display for option in dropdown, otherwise see `getOptionLabel` method. Optional `disabled: boolean` for option to be disabled from selection.
+* `label: ?React$Node`:
+  Component or String representing label for Select component. Default is `''`
+* `caption: ?React$Node`:
+  Component or String representing hint message for Select component. Default is `''`
+* `textValue: ?string`:
+  Initial text value for Search to lookup for text.
+* `error: ?string`:
+  Error message and error state. Default is `''`
+* `multiple: ?boolean`:
+  Sets if multiple choices are allowed in Select component. Default is `false`
+* `autoFocus: boolean`:
+  make the control focused (active). Default is `false`
+* `disabled: boolean`:
+  Disable control from being changed
+* `required: boolean`:
+  Mark control as required
+* `overrides: {}`
+  * `DropDown: ?React$Node` component to use for dropdown list
+  * `Option: ?React$Node` component to use for options in dropdown list
+  * `Root: ?React$Node` component to use for most top of the select component
+  * `Input: ?React$Node` component for Input showing current selected value(s). See `Input` Control of this framework for reference to override it's functionality.
+  * `SearchIcon: ?React$Node` component for all icons appearing in Select component. It's provided `$type: ICON.LOOP | ICON.CLEAR_TAG | ICON.CLEAR_ALL | ICON.SELECTED` to setup corresponding icon of Select component
+  * `Tag: ?React$Node` component for selected options Tags shown in Input for multiple mode selection
+* `onSelect: func(e: SyntheticInputEvent, params: Object)`:
+  handler for events on trigger element when option is selected. `params` has `id` and `label` of selected option and `selectedOptions` array of all of selected.
+* `onUnSelect: func(e: SyntheticInputEvent, params: Object)`:
+  handler for events on trigger element when option is unselected (for multiple choices). `params` has `id` and `label` of unSelected option and `selectedOptions` array of all of selected.
```

Comment:
I think we should just merge these two into a single `onChange(selected: Object | Array<Object>)`. People almost never need the event, so it's pointless to expose it (we can hide it as a second argument if people ask for it). 

And passing people the selected object or array of obejcts (for multiple) makes it easier to store this back in form state.

This also prevents me from needing to add both `onSelect` and `onUnSelect` handlers for multi-select.

react-select uses a single onChange with this signature as well: https://react-select.com/props

Reply:
I have event as first arg, because it's a standard we use across our code so far, but agree it's not needed here.
We do pass all current `selectedOptions` back with it, all in `params` object, but we may still need to know, which one was selected\unselected, that's why we provide it as well.
I can remove UnSelect and Clear handlers, but have to add some extra param of action, like it happens in `react-select` as well